sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -41,10 +59,104 @@ not rely on APIs defined in this specification.*
+
+* Whole graph optimizations become available, including but not limited to:
+** Kernel fusion/fission.
+** Inter-node memory reuse from data staying resident on device.
+** Identification of the peak intermediate output memory requirement, used for
+   more optimal memory allocation.
+
+As well as benefits to the SYCL runtime, there are also advantages to the user
+developing SYCL applications, as repetitive workloads no longer have to
+redundantly issue the same sequence of commands. Instead, a graph is only
+constructed once and submitted for execution as many times as is necessary, only
+changing the data in input buffers or USM allocations. For machine learning
+applications where the same command group pattern is run repeatedly for
+different inputs, this is particularly useful.
+
+=== Requirements
+
+In order to achieve the goals described in previous sections, the following
+requirements were considered:
+
+1. Ability to update inputs/outputs of the graph between submissions, without
+   changing the overall graph structure.
+2. Enable low effort porting of existing applications to use the extension.
+3. Profiling, debugging, and tracing functionality at the granularity of graph
+   nodes.
+4. Integrate sub-graphs (previously constructed graphs) when constructing a new
+   graph.
+5. Support the USM model of memory as well as buffer model.
+6. Compatible with other SYCL extensions and features, e.g kernel fusion &
+   built-in kernels.
+7. Ability to record a graph with commands submitted to different devices in the
+   same context.
+8. A graph constructed using a device queue may be executed on another compatible
+   queue.
+9. Capability to serialize graphs to a binary format which can then be
+   de-serialized and executed. This is helpful for offline cases where a graph
+   can be created by an offline tool to be loaded and run without the end-user
+   incurring the overheads of graph creation.
+10. Backend interoperability, the ability to retrieve a native graph object from
+    the graph and use that in a native backend API.
+
+To allow for prototype implementations of this extension to be developed
+quickly for evaluation the scope of this proposal was limited to a subset
+of these requirements. In particular, the serialization functionality (9),
+backend interoperability (10), and a profiling/debugging interface (3) were
+omitted. As these are not easy to abstract over a number of backends without
+significant investigation. It is also hoped these features can be exposed as
+additive changes to the API, and so in introduced in future versions of the
+extension.
+
+Another reason for deferring a serialize/deserialize API (9) is that its scope
+could extend from emitting the graph in a binary format, to emitting a
+standardized IR format that enables further device specific graph optimizations.
+
+Multi-device support (7) is something we are looking into introducing into
```

Comment:
We should discuss the requirement of allowing an executable graph to be submitted which contains commands targetting different devices. This is not currently possible, but is a feature request we've had from colleagues in Codeplay.

Reply:
Now tracked in https://github.com/reble/llvm/issues/7